src/rapids_metadata/json.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import dataclasses
+import json
+import os
+import sys
+from typing import Any, Union
+
+from . import rapids_metadata
+from .metadata import (
+    RAPIDSMetadata,
+    RAPIDSPackage,
+    RAPIDSRepository,
+    RAPIDSVersion,
+)
+from .rapids_version import get_rapids_version
+
+
+__all__ = [
+    "RAPIDSMetadataEncoder",
+    "main",
+]
+
+
+class RAPIDSMetadataEncoder(json.JSONEncoder):
```

Comment:
What led you down the path of writing a custom JSON encoder?

I just pulled this branch and tested... if we just replace this `PseudoRepository.NVIDIA` ...

https://github.com/rapidsai/rapids-metadata/blob/d3cae20a6be2010a122d7bdecc649f100d0e9a90/src/rapids_metadata/__init__.py#L31

... with a string literal like this:

```python
"__NVIDIA__": RAPIDSRepository(
```

Then the entire thing is JSON-serializable without any customer parsers, like this.

```python
from rapids_metadata import rapids_metadata
from dataclasses import asdict
import json

json.dumps(asdict(rapids_metadata))
```

Would you consider that? Maybe with something like `"__NVIDIA__"` (instead of `"NVIDIA"`) as the literal to make it clearer that it isn't a real `git` repo.

That'd be significantly simpler to read and maintain, without any loss of functionality, as far as I can see.

<details><summary>what that output looks like (click me)</summary>

It produces a single line, but I prettified it for the sake of this comment, so it's easier to visually see the structure.

```json
{
    "versions": {
        "24.08": {
            "repositories": {
                "__no_repo__": {
                    "packages": {
                        "cubinlinker": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "cucim": {
                    "packages": {
                        "cucim": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "cudf": {
                    "packages": {
                        "cudf": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "dask-cudf": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "cugraph": {
                    "packages": {
                        "cugraph": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "cugraph-dgl": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "cugraph-equivariant": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "cugraph-pyg": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "nx-cugraph": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "pylibcugraph": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "cugraph-ops": {
                    "packages": {
                        "pylibcugraphops": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "cuml": {
                    "packages": {
                        "cuml": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "libcuml": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "libcuml-tests": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "cumlprims_mg": {
                    "packages": {
                        "libcumlprims": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "cuproj": {
                    "packages": {
                        "cuproj": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "cuspatial": {
                    "packages": {
                        "cuspatial": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "cuxfilter": {
                    "packages": {
                        "cuxfilter": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "dask-cuda": {
                    "packages": {
                        "dask-cuda": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": false
                        }
                    }
                },
                "ptxcompiler": {
                    "packages": {
                        "ptxcompiler": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "pynvjitlink": {
                    "packages": {
                        "pynvjitlink": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "raft": {
                    "packages": {
                        "libraft": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "libraft-headers": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "pylibraft": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "raft-dask": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "rapids-dask-dependency": {
                    "packages": {
                        "rapids-dask-dependency": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "rmm": {
                    "packages": {
                        "librmm": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "rmm": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "ucxx": {
                    "packages": {
                        "distributed-ucxx": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "libucxx": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        },
                        "ucxx": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "ucx-py": {
                    "packages": {
                        "ucx-py": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                },
                "wholegraph": {
                    "packages": {
                        "pylibwholegraph": {
                            "has_alpha_spec": true,
                            "has_cuda_suffix": true
                        }
                    }
                }
            }
        }
    }
}
```

</details>

Reply:
It was the `PseudoRepository.NVIDIA` that led to the creation of the custom encoder logic. Even if we get rid of that and replace it with simply `"__NVIDIA__"` or `"_nvidia"` (my preference is for the latter), I'd still like to have the `Encoder` class that simply returns `asdict()`.

Reply:
Ok yeah, I don't think the use of that enum justifies maintaining so much encoding logic.

We're not using any of the unique properties of an `Enum` for any functional purpose (e.g. to list a possible set of values to check if a given literal is in a known set of values)... it's just being used as a container for a string.

I think we should replace it with a string literal (whatever one you want is fine), stored in a module-level constant, and switch to using `dataclasses.asdict()`.

> *I'd still like to have the Encoder class that simply returns asdict().*

Sure that's fine, I understand that can be helpful for testing.

Reply:
Done